UnityProject/Assets/Scripts/Input System/CablePlacementVisualisation.cs

```diff
@@ -0,0 +1,239 @@
+	/// <summary>
+	/// color of currentlyHoveredPoint(point below mouse)
+	/// </summary>
+	[SerializeField] private Color onHoverPointColor;
+
+	/// <summary>
+	/// point on which player has pressed mouse button
+	/// </summary>
+	private Vector2Int startPoint;
+	/// <summary>
+	/// point on which player has released mouse button
+	/// </summary>
+	private Vector2Int endPoint;
+	/// <summary>
+	/// point that is currently below mouse
+	/// </summary>
+	private Vector2Int currentlyHoveredPoint;
+
+	/// <summary>
+	/// grid of points
+	/// </summary>
+	private SpriteRenderer[,] connectionPoints = new SpriteRenderer[gridSize, gridSize];
+	/// <summary>
+	/// lineRenderer used to render line between points
+	/// </summary>
+	private LineRenderer lineRenderer;
+
+	/// <summary>
+	/// variable used to restore default color after hover
+	/// </summary>
+	private Color defaultPointColor;
+
+	private void Awake()
+	{
+		// initialize grid
+		for (int y = 0; y < gridSize; y++)
+		{
+			for (int x = 0; x < gridSize; x++)
+			{
+				connectionPoints[x, y] = transform.GetChild(y * gridSize + x).GetComponent<SpriteRenderer>();
+			}
+		}
+
+		// get default color from first point
+		defaultPointColor = connectionPoints[0, 0].color;
+		// get line renderer
+		lineRenderer = GetComponent<LineRenderer>();
+	}
+
+	private void OnEnable()
+	{
+		// reset positions on enable
+		ResetValues();
+	}
+
+	private void Update()
+	{
+		// get releative mouse position
+		Vector2 releativeMousePosition = Camera.main.ScreenToWorldPoint(CommonInput.mousePosition) - transform.position;
+		// get nearest point
+		int x = Mathf.RoundToInt(releativeMousePosition.x * maxIndex);
+		int y = gridSize - 1 - Mathf.RoundToInt(releativeMousePosition.y * maxIndex);
+
+		// clamp to be sure that value is in range <0, maxIndex>
+		x = Mathf.Clamp(x, 0, maxIndex);
+		y = Mathf.Clamp(y, 0, maxIndex);
+		Vector2Int point = new Vector2Int(x, y);
+
+		// if hover point has changed
+		if (point != currentlyHoveredPoint)
+		{
+			// check if current point isn't startPoint or endPoint (to not override color)
+			if (currentlyHoveredPoint != startPoint && currentlyHoveredPoint != endPoint)
+				SetConnectionPointColor(currentlyHoveredPoint.x, currentlyHoveredPoint.y, defaultPointColor);
+
+			SetConnectionPointColor(point.x, point.y, onHoverPointColor);
+			currentlyHoveredPoint = point;
+		}
+
+		// if mouse down - start drawing
+		if (CommonInput.GetMouseButtonDown(0))
+		{
+			startPoint = currentlyHoveredPoint;
+			SetConnectionPointColor(startPoint.x, startPoint.y, startPointColor);
+		}
+		// if mouse up - stop drawing and check if can build
+		else if (CommonInput.GetMouseButtonUp(0))
+		{
+			endPoint = currentlyHoveredPoint;
+			// TODO: check if correct and build
+			Build();
+			ResetValues();
+		}
+
+		// check if startPoint has changed (-Vector2Int.one is default value)
+		if (startPoint != -Vector2Int.one)
+		{
+			lineRenderer.SetPositions(new Vector3[]
```

Comment:
Do we need to set new positions every frame, even if the positions haven't changed? I think we can probably check and only do this as required.

Reply:
fixed (9ef7753f67ad74949d5bfa8329a959a114742c88)